docker/Dockerfile

```diff
@@ -79,12 +72,15 @@ RUN python3 -c "from patchmatch import patch_match"
 FROM python-base AS runtime
 
 # Create a new user
+ARG APPDIR=/usr/src
+ARG APPNAME=InvokeAI
 ARG UNAME=appuser
 RUN useradd \
   --no-log-init \
   -m \
   -U \
-  "${UNAME}"
+  "${UNAME}" \
+  -u 1000
```

Comment:
why do we need to define the UID, isn't it enough to have a user group created with the similar name of the user?

Reply:
Not for podman-- [the way that podman works with rootless containers ](https://www.tutorialworks.com/podman-rootless-volumes/)is that it uses another uid that is distinct from the uid of the user running the container on the host.  In this case, the uid/guid of 1000:1000 in the container is actually something like 100999:100999 in the mounts/volumes.  Or at least that's what you WANT it to be.  So I needed to explicitly set that when the account is created so that the user can be run correctly.  (I could be wrong and there might be another way to do this, but if there is I don't know it.  I could run invokeai in the container as the "root" user rather than as appuser, at which point there would no longer be file access issues, but this seems like a bad idea.)

Reply:
I just built the current main branch locally and executed some commands:

```bash
./docker/build.sh
Activated virtual environment: /Users/mauwii/git/mauwii/InvokeAI/.venv
You are using these values:

Dockerfile:		./Dockerfile
index-url:		https://download.pytorch.org/whl/cpu
Volumename:		invokeai_data
Platform:		linux/arm64
Container Registry:	ghcr.io
Container Repository:	mauwii/invokeai
Container Tag:		main-cpu
Container Flavor:	cpu
Container Image:	ghcr.io/mauwii/invokeai:main-cpu

Volume already exists

[+] Building 178.5s (23/23) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                 0.0s
 => => transferring dockerfile: 2.77kB                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                    0.0s
 => => transferring context: 35B                                                                                                                                                     0.0s
 => resolve image config for docker.io/docker/dockerfile:1                                                                                                                           2.3s
 => [auth] docker/dockerfile:pull token for registry-1.docker.io                                                                                                                     0.0s
 => docker-image://docker.io/docker/dockerfile:1@sha256:39b85bbfa7536a5feceb7372a0817649ecb2724562a38360f4d6a7782a409b14                                                             3.8s
 => => resolve docker.io/docker/dockerfile:1@sha256:39b85bbfa7536a5feceb7372a0817649ecb2724562a38360f4d6a7782a409b14                                                                 0.0s
 => => sha256:39b85bbfa7536a5feceb7372a0817649ecb2724562a38360f4d6a7782a409b14 8.40kB / 8.40kB                                                                                       0.0s
 => => sha256:7f44e51970d0422c2cbff3b20b6b5ef861f6244c396a06e1a96f7aa4fa83a4e6 482B / 482B                                                                                           0.0s
 => => sha256:a28edb2041b8f23c38382d8be273f0239f51ff1f510f98bccc8d0e7f42249e97 2.90kB / 2.90kB                                                                                       0.0s
 => => sha256:9d0cd65540a143ce38aa0be7c5e9efeed30d3580d03667f107cd76354f2bee65 10.82MB / 10.82MB                                                                                     3.1s
 => => extracting sha256:9d0cd65540a143ce38aa0be7c5e9efeed30d3580d03667f107cd76354f2bee65                                                                                            0.6s
 => [internal] load .dockerignore                                                                                                                                                    0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                 0.0s
 => [internal] load metadata for docker.io/library/python:3.9-slim                                                                                                                   0.0s
 => [internal] load build context                                                                                                                                                    0.1s
 => => transferring context: 3.36MB                                                                                                                                                  0.0s
 => [python-base 1/4] FROM docker.io/library/python:3.9-slim                                                                                                                         0.0s
 => CACHED [python-base 2/4] RUN rm -f /etc/apt/apt.conf.d/docker-clean   && echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' >/etc/apt/apt.conf.d/keep-cache               0.0s
 => CACHED [python-base 3/4] RUN   --mount=type=cache,target=/var/cache/apt,sharing=locked   --mount=type=cache,target=/var/lib/apt,sharing=locked   apt-get update   && apt-get in  0.0s
 => CACHED [python-base 4/4] WORKDIR /usr/src                                                                                                                                        0.0s
 => CACHED [pyproject-builder 1/6] RUN   --mount=type=cache,target=/var/cache/apt,sharing=locked   --mount=type=cache,target=/var/lib/apt,sharing=locked   apt-get update   && apt-  0.0s
 => CACHED [pyproject-builder 2/6] RUN mkdir -p /var/cache/buildkit/pip                                                                                                              0.0s
 => CACHED [pyproject-builder 3/6] RUN --mount=type=cache,target=/var/cache/buildkit/pip,sharing=locked   python3 -m venv "InvokeAI"   --upgrade-deps                                0.0s
 => [pyproject-builder 4/6] COPY --link . .                                                                                                                                          0.1s
 => [pyproject-builder 5/6] RUN --mount=type=cache,target=/var/cache/buildkit/pip,sharing=locked   "InvokeAI/bin/pip" install .                                                    148.7s
 => [pyproject-builder 6/6] RUN python3 -c "from patchmatch import patch_match"                                                                                                      3.9s
 => CACHED [runtime 1/3] RUN useradd   --no-log-init   -m   -U   "appuser"                                                                                                           0.0s
 => CACHED [runtime 2/3] RUN mkdir -p "/data"   && chown -R "appuser" "/data"                                                                                                        0.0s
 => [runtime 3/3] COPY --chown=appuser --from=pyproject-builder /usr/src/InvokeAI InvokeAI                                                                                          11.0s
 => exporting to image                                                                                                                                                               4.8s
 => => exporting layers                                                                                                                                                              4.7s
 => => writing image sha256:d274038b0dd470a06f4bcfb8da22fb1fbe071c73ca947d96ef82c5e346dbf62b                                                                                         0.0s
 => => naming to ghcr.io/mauwii/invokeai:main-cpu                                                                                                                                    0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
 ~/git/mauwii/InvokeAI   main ±  docker run --rm --interactive --tty --entrypoint=/bin/bash ghcr.io/mauwii/invokeai:main-cpu
appuser@299ed35c86f9:/usr/src$ id -u
1000
appuser@299ed35c86f9:/usr/src$ id -g
1000
appuser@299ed35c86f9:/usr/src$ whoami
appuser
appuser@299ed35c86f9:/usr/src$ apt-get update
Reading package lists... Done
E: List directory /var/lib/apt/lists/partial is missing. - Acquire (13: Permission denied)
appuser@299ed35c86f9:/usr/src$ sudo apt-get update
bash: sudo: command not found
appuser@299ed35c86f9:/usr/src$
```

- no sudo for building the container
- appuser already has uid 1000
- no sudo inside the container

Reply:
inside the container the uid, guid is 1000.  Same as w/podman.  But if this is rootless docker, try to `touch /data/outputs/testfile` inside the container, and then jump out and look at the uid/guid of the file in ./outputs .  With podman, it's something other than 1000, like some big #.  I believe with rootless docker it's the same, which is why you'd use newuidmap and newguidmap.  Although I don't know much about rootless docker.

Reply:
Also, just out of curiosity, are you running docker run from user 1000/1000?  What happens if you run it rootlessly from 1001/1001?  Will that affect the file ownership of files you create?

Reply:
since /data/outputs is mounted as a bindmount, on my local file system the file permissions are set for my current user (501:20 / mauwii:staff), while in the container they are mounted with permissions set to 1000:1000 / appuser:appuser.

Reply:
Huh... that's very different than rootless podman, where multiple users running in the container would each have their own uid/guid.  If you added a second user, 1001:1001 in the container and created a file, would it still appear as 501:20 outside the container?

Reply:
why should it be different to creating a new user with 1000:1000????

Reply:
@fat-tire is correct, if the user in container is 1000, then in general the files it creates on the bind mounted volume will also be owned by uid=1000. Is it possible that docker on mac changes ownership to the current user for convenience? If so, that's not a standard or generally expected behaviour.